src/arch/x86_64/CodeGen.zig

```diff
@@ -1795,22 +1791,62 @@ fn airStructFieldPtrIndex(self: *Self, inst: Air.Inst.Index, index: u8) !void {
 }
 
 fn structFieldPtr(self: *Self, inst: Air.Inst.Index, operand: Air.Inst.Ref, index: u32) !MCValue {
-    return if (self.liveness.isUnused(inst)) .dead else result: {
-        const mcv = try self.resolveInst(operand);
-        const struct_ty = self.air.typeOf(operand).childType();
-        const struct_size = @intCast(i32, struct_ty.abiSize(self.target.*));
-        const struct_field_offset = @intCast(i32, struct_ty.structFieldOffset(index, self.target.*));
-        const struct_field_ty = struct_ty.structFieldType(index);
-        const struct_field_size = @intCast(i32, struct_field_ty.abiSize(self.target.*));
-
+    if (self.liveness.isUnused(inst)) {
+        return MCValue.dead;
+    }
+    const mcv = try self.resolveInst(operand);
+    const ptr_ty = self.air.typeOf(operand);
+    const struct_ty = ptr_ty.childType();
+    const struct_size = @intCast(u32, struct_ty.abiSize(self.target.*));
+    const struct_field_offset = @intCast(u32, struct_ty.structFieldOffset(index, self.target.*));
+    const struct_field_ty = struct_ty.structFieldType(index);
+    const struct_field_size = @intCast(u32, struct_field_ty.abiSize(self.target.*));
+    const offset_to_field = struct_size - struct_field_offset - struct_field_size;
+
+    const dst_mcv: MCValue = result: {
         switch (mcv) {
+            .stack_offset => {
+                const offset_reg = try self.copyToTmpRegister(ptr_ty, .{
+                    .immediate = offset_to_field,
```

Comment:
Hmm, I think this should be `struct_field_offset`. `mcv` contains the exact run-time address of the struct in memory, so getting a field from that struct should just be `base + offset` operation.

```suggestion
                    .immediate = struct_field offset,
```

Reply:
Hmm, right, lemme double check with multiple struct fields.

Reply:
You were absolutely right, thanks @joachimschmidt557. While at it, I've also fixed how we lower the struct to bytes in the final binary - prior to cf503a5 we would not correctly save each field at its required alignment leading to wrong field accesses in the generated machine code.